File: kattis/cats.py
#TLE on Python solution, accepted on C++ solution. But the idea is totally the same.

class Group:
	def __init__(self,_id,level=0,parent=None):
		self.id = _id
		self.level = level
		self.parent = parent
	def traceTop(self):
		cur = self
		while cur.parent:
			cur = cur.parent
		return cur
	def sameGroup(self,other):
		return self.traceTop()==other.traceTop()
	def join(self,other,maxId):
		if not self.sameGroup(other):
			left,right = self.traceTop(),other.traceTop()
			if left.level>right.level:
				right.parent = left
			elif left.level<right.level:
				left.parent = right
			else:
				newParent = Group(maxId,left.level+1,None)
				left.parent,right.parent = newParent,newParent
				maxId += 1
		return maxId


def mstLen(edges,n):
	groups,curId,res = [Group(i) for i in xrange(n)],n,0
	for d,left,right in edges:
		if not groups[left].sameGroup(groups[right]):
			curId = groups[left].join(groups[right],curId)
			res += d
	return res

tt = int(raw_input())
for t in xrange(tt):
	m,c = map(int,raw_input().split())
	edges = []
	for i in xrange(c*(c-1)/2):
		left,right,d = map(int,raw_input().split())
		edges.append((d,left,right))
	edges.sort()
	spill = mstLen(edges,c)+c
	if spill<=m:
		print "yes"
	else:
		print "no"

File: kattis/cats.cc
#include<iostream>
#include<vector>
#include<tuple>
#include<algorithm>
using namespace std;

class Group{
public:
	int id,level;
	Group *top = NULL;
	Group(int id,int level=0):id(id),level(level){}
	Group* traceTop(){
		Group *cur = this;
		while(cur->top){
			cur = cur->top;
		}
		return cur;
	}
	void setParent(Group *other){
		top = other;
	}
	int sameGroup(Group &other){
		if(traceTop()==other.traceTop())return 1;
		return 0;
	}
	int join(Group &other,int maxId){
		if(!sameGroup(other)){
			Group *left=traceTop(),*right = other.traceTop();
			if(left->level > right->level){
				right->setParent(left);
			}else if(left->level < right->level){
				left->setParent(right);
			}else{
				Group *newTop = new Group(maxId++,level+1);
				left->setParent(newTop);
				right->setParent(newTop);
			}
		}
		return maxId;
	}
};

int mstLen(vector<tuple<int,int,int> > &edges,int n){
	vector<Group> groups;
	int ans = 0,curId = n;
	for(int i=0;i<n;i++){
		Group g(i);
		groups.push_back(g);
	}
	for(auto it=edges.begin();it!=edges.end();it++){
		int d,left,right;
		tie(d,left,right) = (*it);
		if(!groups[left].sameGroup(groups[right])){
			ans += d;
			curId = groups[left].join(groups[right],curId);
		}
	}
	return ans;
}

int main(){
	int tt;cin>>tt;
	for(int i=0;i<tt;i++){
		int m,c;cin>>m>>c;
		vector<tuple<int,int,int> > edges;
		for(int j=0;j<c*(c-1)/2;j++){
			int left,right,d;
			cin>>left>>right>>d;
			edges.push_back(make_tuple(d,left,right));
		}
		sort(edges.begin(),edges.end());
		int spill = mstLen(edges,c) + c;
		if(spill<=m)cout<<"yes"<<endl;
		else cout<<"no"<<endl;
	}
	return 0;
}















